{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.Seq import Seq\n",
    "from Bio import pairwise2"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 双序列比对\n",
    "\n",
    "\n",
    "对于Biopython而言\n",
    "\n",
    "The names of the alignment functions in this module follow the\n",
    "    convention\n",
    "    <alignment type>XX\n",
    "    where <alignment type> is either \"global\" or \"local\" and XX is a 2\n",
    "    character code indicating the parameters it takes.  The first\n",
    "    character indicates the parameters for matches (and mismatches), and\n",
    "    the second indicates the parameters for gap penalties.\n",
    "    \n",
    "    The match parameters are::\n",
    "    \n",
    "        CODE  DESCRIPTION\n",
    "        x     No parameters. Identical characters have score of 1, otherwise 0. match为1分，mismatch为0分\n",
    "        m     A match score is the score of identical chars, otherwise mismatch\n",
    "              score.\n",
    "        d     A dictionary returns the score of any pair of characters.\n",
    "        c     A callback function returns scores.\n",
    "    \n",
    "    The gap penalty parameters are::\n",
    "    \n",
    "        CODE  DESCRIPTION\n",
    "        x     No gap penalties. 没有空位罚分\n",
    "        s     Same open and extend gap penalties for both sequences.\n",
    "        d     The sequences have different open and extend gap penalties.\n",
    "        c     A callback function returns the gap penalties.\n",
    "    \n",
    "    All the different alignment functions are contained in an object\n",
    "    ``align``. For example:\n",
    "    \n",
    "    \n",
    "虽然这么多可以使用得比对函数。\n",
    "但是一般常用得是以下\n",
    "\n",
    "- `globalxx(seq1,seq2)` match为1分，mismatch为0分，空位罚分为0 \n",
    "- `globalms(seq1,seq2,2,-1,-0.5,-0.1)`  指定match为2分，错配罚分为1分，空位罚分0.5，延伸空位罚分0.1\n",
    "- `globaldx(seq1,seq2,matrix)` 通过指定打分矩阵，一般是蛋白质序列得比对采用。\n",
    "\n",
    "对于不知道如何使用得 直接调用`help(pairwise2.align.globaldc)`来查询相应得使用帮助"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "seq1 = Seq('ACACACTA')\n",
    "seq2 =Seq('AGCACACA')"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### gloablxx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "alignments = pairwise2.align.globalxx(seq1,seq2)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A-CACACTA\n",
      "| ||||| |\n",
      "AGCACAC-A\n",
      "  Score=7\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for a in pairwise2.align.globalxx(seq1, seq2):\n",
    "    print(format_alignment(*a))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "使用`format_alignment(*alignments[0])`可以格式化显示比对结果\n",
    "但是仅仅会显示匹配得部分，**不会显示全部序列**"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A-CACACTA\n",
      "| ||||| |\n",
      "AGCACAC-A\n",
      "  Score=7\n",
      "\n"
     ]
    }
   ],
   "source": [
    "from Bio.pairwise2 import format_alignment\n",
    "print(format_alignment(*alignments[0]))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### globalms"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "如果要指定相关得罚分则需要使用\n",
    "\n",
    "指定match为2分，错配罚分为1分，空位罚分0.5，延伸空位罚分0.1\n",
    "\n",
    "`pairwise2.align.globalms(seq1,seq2,2,1,-0.5,-0.1)`"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A-CACACTA\n",
      "| ||||| |\n",
      "AGCACAC-A\n",
      "  Score=13\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for a in pairwise2.align.globalms(seq1,seq2,2,-1,-0.5,-0.1):\n",
    "    print(format_alignment(*a)) "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### globaldx"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "KEVLA\n",
      " ||| \n",
      "-EVL-\n",
      "  Score=22\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#导入打分矩阵相关得信息\n",
    "from Bio.SubsMat import MatrixInfo as matlist\n",
    "#选择pam30作为打分矩阵\n",
    "PAM30= matlist.pam30\n",
    "\n",
    "for a in pairwise2.align.globaldx('KEVLA','EVL',PAM30):\n",
    "    print(format_alignment(*a,full_sequences=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### 局部比对\n",
    "与全局比对没区别。\n",
    "\n",
    "注意gloablxx 改为localxx即可。"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "1 ACCG\n",
      "  | ||\n",
      "1 A-CG\n",
      "  Score=3\n",
      "\n",
      "1 ACCG\n",
      "  || |\n",
      "1 AC-G\n",
      "  Score=3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "alignments = pairwise2.align.localxx('ACCGT','ACG')\n",
    "for i in alignments:\n",
    "    print(format_alignment(*i))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "给format_alignment()指定位置参数 full_sequences =True即可显示完整得序列"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "ACCGT\n",
      "| || \n",
      "A-CG-\n",
      "  Score=3\n",
      "\n",
      "ACCGT\n",
      "|| | \n",
      "AC-G-\n",
      "  Score=3\n",
      "\n"
     ]
    }
   ],
   "source": [
    "for i in alignments:\n",
    "    print(format_alignment(*i,full_sequences=True))"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "##### 打分矩阵\n",
    "\n",
    "`from Bio.SubsMat import MatrixInfo as matlist`导入矩阵相关得信息， 从中选择需要得矩阵即可"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio.SubsMat import MatrixInfo as matlist"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {
    "collapsed": true
   },
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{('W', 'F'): -4,\n",
       " ('L', 'R'): -8,\n",
       " ('S', 'P'): -2,\n",
       " ('V', 'T'): -3,\n",
       " ('Q', 'Q'): 8,\n",
       " ('N', 'A'): -4,\n",
       " ('Z', 'Y'): -9,\n",
       " ('W', 'R'): -2,\n",
       " ('Q', 'A'): -4,\n",
       " ('S', 'D'): -4,\n",
       " ('H', 'H'): 9,\n",
       " ('S', 'H'): -6,\n",
       " ('H', 'D'): -4,\n",
       " ('L', 'N'): -7,\n",
       " ('W', 'A'): -13,\n",
       " ('Y', 'M'): -11,\n",
       " ('G', 'R'): -9,\n",
       " ('Y', 'I'): -6,\n",
       " ('Y', 'E'): -8,\n",
       " ('B', 'Y'): -6,\n",
       " ('Y', 'A'): -8,\n",
       " ('V', 'D'): -8,\n",
       " ('B', 'S'): -1,\n",
       " ('Y', 'Y'): 10,\n",
       " ('G', 'N'): -3,\n",
       " ('E', 'C'): -14,\n",
       " ('Y', 'Q'): -12,\n",
       " ('Z', 'Z'): 6,\n",
       " ('V', 'A'): -2,\n",
       " ('C', 'C'): 10,\n",
       " ('M', 'R'): -4,\n",
       " ('V', 'E'): -6,\n",
       " ('T', 'N'): -2,\n",
       " ('P', 'P'): 8,\n",
       " ('V', 'I'): 2,\n",
       " ('V', 'S'): -6,\n",
       " ('Z', 'P'): -4,\n",
       " ('V', 'M'): -1,\n",
       " ('T', 'F'): -9,\n",
       " ('V', 'Q'): -7,\n",
       " ('K', 'K'): 7,\n",
       " ('P', 'D'): -8,\n",
       " ('I', 'H'): -9,\n",
       " ('I', 'D'): -7,\n",
       " ('T', 'R'): -6,\n",
       " ('P', 'L'): -7,\n",
       " ('K', 'G'): -7,\n",
       " ('M', 'N'): -9,\n",
       " ('P', 'H'): -4,\n",
       " ('F', 'Q'): -13,\n",
       " ('Z', 'G'): -5,\n",
       " ('X', 'L'): -6,\n",
       " ('T', 'M'): -4,\n",
       " ('Z', 'C'): -14,\n",
       " ('X', 'H'): -5,\n",
       " ('D', 'R'): -10,\n",
       " ('B', 'W'): -10,\n",
       " ('X', 'D'): -5,\n",
       " ('Z', 'K'): -4,\n",
       " ('F', 'A'): -8,\n",
       " ('Z', 'W'): -14,\n",
       " ('F', 'E'): -14,\n",
       " ('D', 'N'): 2,\n",
       " ('B', 'K'): -2,\n",
       " ('X', 'X'): -5,\n",
       " ('F', 'I'): -2,\n",
       " ('B', 'G'): -3,\n",
       " ('X', 'T'): -4,\n",
       " ('F', 'M'): -4,\n",
       " ('B', 'C'): -12,\n",
       " ('Z', 'I'): -6,\n",
       " ('Z', 'V'): -6,\n",
       " ('S', 'S'): 6,\n",
       " ('L', 'Q'): -5,\n",
       " ('W', 'E'): -17,\n",
       " ('Q', 'R'): -2,\n",
       " ('N', 'N'): 8,\n",
       " ('W', 'M'): -13,\n",
       " ('Q', 'C'): -14,\n",
       " ('W', 'I'): -14,\n",
       " ('S', 'C'): -3,\n",
       " ('L', 'A'): -6,\n",
       " ('S', 'G'): -2,\n",
       " ('L', 'E'): -9,\n",
       " ('W', 'Q'): -13,\n",
       " ('H', 'G'): -9,\n",
       " ('S', 'K'): -4,\n",
       " ('Q', 'N'): -3,\n",
       " ('N', 'R'): -6,\n",
       " ('H', 'C'): -7,\n",
       " ('Y', 'N'): -4,\n",
       " ('G', 'Q'): -7,\n",
       " ('Y', 'F'): 2,\n",
       " ('C', 'A'): -6,\n",
       " ('V', 'L'): -2,\n",
       " ('G', 'E'): -4,\n",
       " ('G', 'A'): -2,\n",
       " ('K', 'R'): 0,\n",
       " ('E', 'D'): 2,\n",
       " ('Y', 'R'): -10,\n",
       " ('M', 'Q'): -4,\n",
       " ('T', 'I'): -2,\n",
       " ('C', 'D'): -14,\n",
       " ('V', 'F'): -8,\n",
       " ('T', 'A'): -1,\n",
       " ('T', 'P'): -4,\n",
       " ('B', 'P'): -7,\n",
       " ('T', 'E'): -6,\n",
       " ('V', 'N'): -8,\n",
       " ('P', 'G'): -6,\n",
       " ('M', 'A'): -5,\n",
       " ('K', 'H'): -6,\n",
       " ('V', 'R'): -8,\n",
       " ('P', 'C'): -8,\n",
       " ('M', 'E'): -7,\n",
       " ('K', 'L'): -8,\n",
       " ('V', 'V'): 7,\n",
       " ('M', 'I'): -1,\n",
       " ('T', 'Q'): -5,\n",
       " ('I', 'G'): -11,\n",
       " ('P', 'K'): -6,\n",
       " ('M', 'M'): 11,\n",
       " ('K', 'D'): -4,\n",
       " ('I', 'C'): -6,\n",
       " ('Z', 'D'): 1,\n",
       " ('F', 'R'): -9,\n",
       " ('X', 'K'): -5,\n",
       " ('Q', 'D'): -2,\n",
       " ('X', 'G'): -5,\n",
       " ('Z', 'L'): -7,\n",
       " ('X', 'C'): -9,\n",
       " ('Z', 'H'): -1,\n",
       " ('B', 'L'): -9,\n",
       " ('B', 'H'): -1,\n",
       " ('F', 'F'): 9,\n",
       " ('X', 'W'): -11,\n",
       " ('B', 'D'): 6,\n",
       " ('D', 'A'): -3,\n",
       " ('S', 'L'): -8,\n",
       " ('X', 'S'): -3,\n",
       " ('F', 'N'): -9,\n",
       " ('S', 'R'): -3,\n",
       " ('W', 'D'): -15,\n",
       " ('V', 'Y'): -7,\n",
       " ('W', 'L'): -6,\n",
       " ('H', 'R'): -2,\n",
       " ('W', 'H'): -7,\n",
       " ('H', 'N'): 0,\n",
       " ('W', 'T'): -13,\n",
       " ('T', 'T'): 7,\n",
       " ('S', 'F'): -6,\n",
       " ('W', 'P'): -14,\n",
       " ('L', 'D'): -12,\n",
       " ('B', 'I'): -6,\n",
       " ('L', 'H'): -6,\n",
       " ('S', 'N'): 0,\n",
       " ('B', 'T'): -3,\n",
       " ('L', 'L'): 7,\n",
       " ('Y', 'K'): -9,\n",
       " ('E', 'Q'): 1,\n",
       " ('Y', 'G'): -14,\n",
       " ('Z', 'S'): -5,\n",
       " ('Y', 'C'): -4,\n",
       " ('G', 'D'): -3,\n",
       " ('B', 'V'): -8,\n",
       " ('E', 'A'): -2,\n",
       " ('Y', 'W'): -5,\n",
       " ('E', 'E'): 8,\n",
       " ('Y', 'S'): -7,\n",
       " ('C', 'N'): -11,\n",
       " ('V', 'C'): -6,\n",
       " ('T', 'H'): -7,\n",
       " ('P', 'R'): -4,\n",
       " ('V', 'G'): -5,\n",
       " ('T', 'L'): -7,\n",
       " ('V', 'K'): -9,\n",
       " ('K', 'Q'): -3,\n",
       " ('R', 'A'): -7,\n",
       " ('I', 'R'): -5,\n",
       " ('T', 'D'): -5,\n",
       " ('P', 'F'): -10,\n",
       " ('I', 'N'): -5,\n",
       " ('K', 'I'): -6,\n",
       " ('M', 'D'): -11,\n",
       " ('V', 'W'): -15,\n",
       " ('W', 'W'): 13,\n",
       " ('M', 'H'): -10,\n",
       " ('P', 'N'): -6,\n",
       " ('K', 'A'): -7,\n",
       " ('M', 'L'): 1,\n",
       " ('K', 'E'): -4,\n",
       " ('Z', 'E'): 6,\n",
       " ('X', 'N'): -3,\n",
       " ('Z', 'A'): -3,\n",
       " ('Z', 'M'): -5,\n",
       " ('X', 'F'): -8,\n",
       " ('K', 'C'): -14,\n",
       " ('B', 'Q'): -3,\n",
       " ('X', 'B'): -5,\n",
       " ('B', 'M'): -10,\n",
       " ('F', 'C'): -13,\n",
       " ('Z', 'Q'): 6,\n",
       " ('X', 'Z'): -5,\n",
       " ('F', 'G'): -9,\n",
       " ('B', 'E'): 1,\n",
       " ('X', 'V'): -5,\n",
       " ('F', 'K'): -14,\n",
       " ('B', 'A'): -3,\n",
       " ('X', 'R'): -6,\n",
       " ('D', 'D'): 8,\n",
       " ('W', 'G'): -15,\n",
       " ('Z', 'F'): -13,\n",
       " ('S', 'Q'): -5,\n",
       " ('W', 'C'): -15,\n",
       " ('W', 'K'): -12,\n",
       " ('H', 'Q'): 1,\n",
       " ('L', 'C'): -15,\n",
       " ('W', 'N'): -8,\n",
       " ('S', 'A'): 0,\n",
       " ('L', 'G'): -10,\n",
       " ('W', 'S'): -5,\n",
       " ('S', 'E'): -4,\n",
       " ('H', 'E'): -5,\n",
       " ('S', 'I'): -7,\n",
       " ('H', 'A'): -7,\n",
       " ('S', 'M'): -5,\n",
       " ('Y', 'L'): -7,\n",
       " ('Y', 'H'): -3,\n",
       " ('Y', 'D'): -11,\n",
       " ('E', 'R'): -9,\n",
       " ('X', 'P'): -5,\n",
       " ('G', 'G'): 6,\n",
       " ('G', 'C'): -9,\n",
       " ('E', 'N'): -2,\n",
       " ('Y', 'T'): -6,\n",
       " ('Y', 'P'): -13,\n",
       " ('T', 'K'): -3,\n",
       " ('A', 'A'): 6,\n",
       " ('P', 'Q'): -3,\n",
       " ('T', 'C'): -8,\n",
       " ('V', 'H'): -6,\n",
       " ('T', 'G'): -6,\n",
       " ('I', 'Q'): -8,\n",
       " ('Z', 'T'): -6,\n",
       " ('C', 'R'): -8,\n",
       " ('V', 'P'): -6,\n",
       " ('P', 'E'): -5,\n",
       " ('M', 'C'): -13,\n",
       " ('K', 'N'): -1,\n",
       " ('I', 'I'): 8,\n",
       " ('P', 'A'): -2,\n",
       " ('M', 'G'): -8,\n",
       " ('T', 'S'): 0,\n",
       " ('I', 'E'): -5,\n",
       " ('P', 'M'): -8,\n",
       " ('M', 'K'): -2,\n",
       " ('I', 'A'): -5,\n",
       " ('P', 'I'): -8,\n",
       " ('R', 'R'): 8,\n",
       " ('X', 'M'): -5,\n",
       " ('L', 'I'): -1,\n",
       " ('X', 'I'): -5,\n",
       " ('Z', 'B'): 0,\n",
       " ('X', 'E'): -5,\n",
       " ('Z', 'N'): -3,\n",
       " ('X', 'A'): -3,\n",
       " ('B', 'R'): -7,\n",
       " ('B', 'N'): 6,\n",
       " ('F', 'D'): -15,\n",
       " ('X', 'Y'): -7,\n",
       " ('Z', 'R'): -4,\n",
       " ('F', 'H'): -6,\n",
       " ('B', 'F'): -10,\n",
       " ('F', 'L'): -3,\n",
       " ('X', 'Q'): -5,\n",
       " ('B', 'B'): 6}"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#选择pam30\n",
    "PAM30= matlist.pam30\n",
    "PAM30"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "from Bio import Align "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Help on package Bio.Align in Bio:\n",
      "\n",
      "NAME\n",
      "    Bio.Align - Code for dealing with sequence alignments.\n",
      "\n",
      "DESCRIPTION\n",
      "    One of the most important things in this module is the MultipleSeqAlignment\n",
      "    class, used in the Bio.AlignIO module.\n",
      "\n",
      "PACKAGE CONTENTS\n",
      "    AlignInfo\n",
      "    Applications (package)\n",
      "    _aligners\n",
      "    substitution_matrices (package)\n",
      "\n",
      "CLASSES\n",
      "    _algorithms.PairwiseAligner(builtins.object)\n",
      "        PairwiseAligner\n",
      "    builtins.object\n",
      "        MultipleSeqAlignment\n",
      "        PairwiseAlignment\n",
      "        PairwiseAlignments\n",
      "    \n",
      "    class MultipleSeqAlignment(builtins.object)\n",
      "     |  MultipleSeqAlignment(records, alphabet=None, annotations=None, column_annotations=None)\n",
      "     |  \n",
      "     |  Represents a classical multiple sequence alignment (MSA).\n",
      "     |  \n",
      "     |  By this we mean a collection of sequences (usually shown as rows) which\n",
      "     |  are all the same length (usually with gap characters for insertions or\n",
      "     |  padding). The data can then be regarded as a matrix of letters, with well\n",
      "     |  defined columns.\n",
      "     |  \n",
      "     |  You would typically create an MSA by loading an alignment file with the\n",
      "     |  AlignIO module:\n",
      "     |  \n",
      "     |  >>> from Bio import AlignIO\n",
      "     |  >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
      "     |  >>> print(align)\n",
      "     |  SingleLetterAlphabet() alignment with 7 rows and 156 columns\n",
      "     |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |  TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |  TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |  TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  In some respects you can treat these objects as lists of SeqRecord objects,\n",
      "     |  each representing a row of the alignment. Iterating over an alignment gives\n",
      "     |  the SeqRecord object for each row:\n",
      "     |  \n",
      "     |  >>> len(align)\n",
      "     |  7\n",
      "     |  >>> for record in align:\n",
      "     |  ...     print(\"%s %i\" % (record.id, len(record)))\n",
      "     |  ...\n",
      "     |  gi|6273285|gb|AF191659.1|AF191 156\n",
      "     |  gi|6273284|gb|AF191658.1|AF191 156\n",
      "     |  gi|6273287|gb|AF191661.1|AF191 156\n",
      "     |  gi|6273286|gb|AF191660.1|AF191 156\n",
      "     |  gi|6273290|gb|AF191664.1|AF191 156\n",
      "     |  gi|6273289|gb|AF191663.1|AF191 156\n",
      "     |  gi|6273291|gb|AF191665.1|AF191 156\n",
      "     |  \n",
      "     |  You can also access individual rows as SeqRecord objects via their index:\n",
      "     |  \n",
      "     |  >>> print(align[0].id)\n",
      "     |  gi|6273285|gb|AF191659.1|AF191\n",
      "     |  >>> print(align[-1].id)\n",
      "     |  gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  And extract columns as strings:\n",
      "     |  \n",
      "     |  >>> print(align[:, 1])\n",
      "     |  AAAAAAA\n",
      "     |  \n",
      "     |  Or, take just the first ten columns as a sub-alignment:\n",
      "     |  \n",
      "     |  >>> print(align[:, :10])\n",
      "     |  SingleLetterAlphabet() alignment with 7 rows and 10 columns\n",
      "     |  TATACATTAA gi|6273285|gb|AF191659.1|AF191\n",
      "     |  TATACATTAA gi|6273284|gb|AF191658.1|AF191\n",
      "     |  TATACATTAA gi|6273287|gb|AF191661.1|AF191\n",
      "     |  TATACATAAA gi|6273286|gb|AF191660.1|AF191\n",
      "     |  TATACATTAA gi|6273290|gb|AF191664.1|AF191\n",
      "     |  TATACATTAA gi|6273289|gb|AF191663.1|AF191\n",
      "     |  TATACATTAA gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  Combining this alignment slicing with alignment addition allows you to\n",
      "     |  remove a section of the alignment. For example, taking just the first\n",
      "     |  and last ten columns:\n",
      "     |  \n",
      "     |  >>> print(align[:, :10] + align[:, -10:])\n",
      "     |  SingleLetterAlphabet() alignment with 7 rows and 20 columns\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |  TATACATAAAGTGTACCAGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |  TATACATTAAGTATACCAGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |  TATACATTAAGTGTACCAGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |  \n",
      "     |  Note - This object replaced the older Alignment object defined in module\n",
      "     |  Bio.Align.Generic but is not fully backwards compatible with it.\n",
      "     |  \n",
      "     |  Note - This object does NOT attempt to model the kind of alignments used\n",
      "     |  in next generation sequencing with multiple sequencing reads which are\n",
      "     |  much shorter than the alignment, and where there is usually a consensus or\n",
      "     |  reference sequence with special status.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __add__(self, other)\n",
      "     |      Combine two alignments with the same number of rows by adding them.\n",
      "     |      \n",
      "     |      If you have two multiple sequence alignments (MSAs), there are two ways to think\n",
      "     |      about adding them - by row or by column. Using the extend method adds by row.\n",
      "     |      Using the addition operator adds by column. For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import generic_dna\n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a1 = SeqRecord(Seq(\"AAAAC\", generic_dna), id=\"Alpha\")\n",
      "     |      >>> b1 = SeqRecord(Seq(\"AAA-C\", generic_dna), id=\"Beta\")\n",
      "     |      >>> c1 = SeqRecord(Seq(\"AAAAG\", generic_dna), id=\"Gamma\")\n",
      "     |      >>> a2 = SeqRecord(Seq(\"GT\", generic_dna), id=\"Alpha\")\n",
      "     |      >>> b2 = SeqRecord(Seq(\"GT\", generic_dna), id=\"Beta\")\n",
      "     |      >>> c2 = SeqRecord(Seq(\"GT\", generic_dna), id=\"Gamma\")\n",
      "     |      >>> left = MultipleSeqAlignment([a1, b1, c1],\n",
      "     |      ...                             annotations={\"tool\": \"demo\", \"name\": \"start\"},\n",
      "     |      ...                             column_annotations={\"stats\": \"CCCXC\"})\n",
      "     |      >>> right = MultipleSeqAlignment([a2, b2, c2],\n",
      "     |      ...                             annotations={\"tool\": \"demo\", \"name\": \"end\"},\n",
      "     |      ...                             column_annotations={\"stats\": \"CC\"})\n",
      "     |      \n",
      "     |      Now, let's look at these two alignments:\n",
      "     |      \n",
      "     |      >>> print(left)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 5 columns\n",
      "     |      AAAAC Alpha\n",
      "     |      AAA-C Beta\n",
      "     |      AAAAG Gamma\n",
      "     |      >>> print(right)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 2 columns\n",
      "     |      GT Alpha\n",
      "     |      GT Beta\n",
      "     |      GT Gamma\n",
      "     |      \n",
      "     |      And add them:\n",
      "     |      \n",
      "     |      >>> combined = left + right\n",
      "     |      >>> print(combined)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      \n",
      "     |      For this to work, both alignments must have the same number of records (here\n",
      "     |      they both have 3 rows):\n",
      "     |      \n",
      "     |      >>> len(left)\n",
      "     |      3\n",
      "     |      >>> len(right)\n",
      "     |      3\n",
      "     |      >>> len(combined)\n",
      "     |      3\n",
      "     |      \n",
      "     |      The individual rows are SeqRecord objects, and these can be added together. Refer\n",
      "     |      to the SeqRecord documentation for details of how the annotation is handled. This\n",
      "     |      example is a special case in that both original alignments shared the same names,\n",
      "     |      meaning when the rows are added they also get the same name.\n",
      "     |      \n",
      "     |      Any common annotations are preserved, but differing annotation is lost. This is\n",
      "     |      the same behaviour used in the SeqRecord annotations and is designed to prevent\n",
      "     |      accidental propagation of inappropriate values:\n",
      "     |      \n",
      "     |      >>> combined.annotations\n",
      "     |      {'tool': 'demo'}\n",
      "     |      \n",
      "     |      Similarly any common per-column-annotations are combined:\n",
      "     |      \n",
      "     |      >>> combined.column_annotations\n",
      "     |      {'stats': 'CCCXCCC'}\n",
      "     |  \n",
      "     |  __format__(self, format_spec)\n",
      "     |      Return the alignment as a string in the specified file format.\n",
      "     |      \n",
      "     |      This method supports the python format() function added in\n",
      "     |      Python 2.6/3.0.  The format_spec should be a lower case\n",
      "     |      string supported by Bio.AlignIO as an output file format.\n",
      "     |      See also the alignment's format() method.\n",
      "     |  \n",
      "     |  __getitem__(self, index)\n",
      "     |      Access part of the alignment.\n",
      "     |      \n",
      "     |      Depending on the indices, you can get a SeqRecord object\n",
      "     |      (representing a single row), a Seq object (for a single columns),\n",
      "     |      a string (for a single characters) or another alignment\n",
      "     |      (representing some part or all of the alignment).\n",
      "     |      \n",
      "     |      align[r,c] gives a single character as a string\n",
      "     |      align[r] gives a row as a SeqRecord\n",
      "     |      align[r,:] gives a row as a SeqRecord\n",
      "     |      align[:,c] gives a column as a Seq (using the alignment's alphabet)\n",
      "     |      \n",
      "     |      align[:] and align[:,:] give a copy of the alignment\n",
      "     |      \n",
      "     |      Anything else gives a sub alignment, e.g.\n",
      "     |      align[0:2] or align[0:2,:] uses only row 0 and 1\n",
      "     |      align[:,1:3] uses only columns 1 and 2\n",
      "     |      align[0:2,1:3] uses only rows 0 & 1 and only cols 1 & 2\n",
      "     |      \n",
      "     |      We'll use the following example alignment here for illustration:\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import generic_dna\n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"AAA-CGT\", generic_dna), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"AAAAGGT\", generic_dna), id=\"Gamma\")\n",
      "     |      >>> d = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Delta\")\n",
      "     |      >>> e = SeqRecord(Seq(\"AAA-GGT\", generic_dna), id=\"Epsilon\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c, d, e], generic_dna)\n",
      "     |      \n",
      "     |      You can access a row of the alignment as a SeqRecord using an integer\n",
      "     |      index (think of the alignment as a list of SeqRecord objects here):\n",
      "     |      \n",
      "     |      >>> first_record = align[0]\n",
      "     |      >>> print(\"%s %s\" % (first_record.id, first_record.seq))\n",
      "     |      Alpha AAAACGT\n",
      "     |      >>> last_record = align[-1]\n",
      "     |      >>> print(\"%s %s\" % (last_record.id, last_record.seq))\n",
      "     |      Epsilon AAA-GGT\n",
      "     |      \n",
      "     |      You can also access use python's slice notation to create a sub-alignment\n",
      "     |      containing only some of the SeqRecord objects:\n",
      "     |      \n",
      "     |      >>> sub_alignment = align[2:5]\n",
      "     |      >>> print(sub_alignment)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 7 columns\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAAACGT Delta\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      \n",
      "     |      This includes support for a step, i.e. align[start:end:step], which\n",
      "     |      can be used to select every second sequence:\n",
      "     |      \n",
      "     |      >>> sub_alignment = align[::2]\n",
      "     |      >>> print(sub_alignment)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      \n",
      "     |      Or to get a copy of the alignment with the rows in reverse order:\n",
      "     |      \n",
      "     |      >>> rev_alignment = align[::-1]\n",
      "     |      >>> print(rev_alignment)\n",
      "     |      DNAAlphabet() alignment with 5 rows and 7 columns\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      AAAACGT Delta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAACGT Alpha\n",
      "     |      \n",
      "     |      You can also use two indices to specify both rows and columns. Using simple\n",
      "     |      integers gives you the entry as a single character string. e.g.\n",
      "     |      \n",
      "     |      >>> align[3, 4]\n",
      "     |      'C'\n",
      "     |      \n",
      "     |      This is equivalent to:\n",
      "     |      \n",
      "     |      >>> align[3][4]\n",
      "     |      'C'\n",
      "     |      \n",
      "     |      or:\n",
      "     |      \n",
      "     |      >>> align[3].seq[4]\n",
      "     |      'C'\n",
      "     |      \n",
      "     |      To get a single column (as a string) use this syntax:\n",
      "     |      \n",
      "     |      >>> align[:, 4]\n",
      "     |      'CCGCG'\n",
      "     |      \n",
      "     |      Or, to get part of a column,\n",
      "     |      \n",
      "     |      >>> align[1:3, 4]\n",
      "     |      'CG'\n",
      "     |      \n",
      "     |      However, in general you get a sub-alignment,\n",
      "     |      \n",
      "     |      >>> print(align[1:5, 3:6])\n",
      "     |      DNAAlphabet() alignment with 4 rows and 3 columns\n",
      "     |      -CG Beta\n",
      "     |      AGG Gamma\n",
      "     |      ACG Delta\n",
      "     |      -GG Epsilon\n",
      "     |      \n",
      "     |      This should all seem familiar to anyone who has used the NumPy\n",
      "     |      array or matrix objects.\n",
      "     |  \n",
      "     |  __init__(self, records, alphabet=None, annotations=None, column_annotations=None)\n",
      "     |      Initialize a new MultipleSeqAlignment object.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - records - A list (or iterator) of SeqRecord objects, whose\n",
      "     |                   sequences are all the same length.  This may be an be an\n",
      "     |                   empty list.\n",
      "     |       - alphabet - The alphabet for the whole alignment, typically a gapped\n",
      "     |                    alphabet, which should be a super-set of the individual\n",
      "     |                    record alphabets.  If omitted, a consensus alphabet is\n",
      "     |                    used.\n",
      "     |       - annotations - Information about the whole alignment (dictionary).\n",
      "     |       - column_annotations - Per column annotation (restricted dictionary).\n",
      "     |                    This holds Python sequences (lists, strings, tuples)\n",
      "     |                    whose length matches the number of columns. A typical\n",
      "     |                    use would be a secondary structure consensus string.\n",
      "     |      \n",
      "     |      You would normally load a MSA from a file using Bio.AlignIO, but you\n",
      "     |      can do this from a list of SeqRecord objects too:\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import generic_dna\n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"AAA-CGT\", generic_dna), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"AAAAGGT\", generic_dna), id=\"Gamma\")\n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c],\n",
      "     |      ...                              annotations={\"tool\": \"demo\"},\n",
      "     |      ...                              column_annotations={\"stats\": \"CCCXCCC\"})\n",
      "     |      >>> print(align)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      >>> align.annotations\n",
      "     |      {'tool': 'demo'}\n",
      "     |      >>> align.column_annotations\n",
      "     |      {'stats': 'CCCXCCC'}\n",
      "     |  \n",
      "     |  __iter__(self)\n",
      "     |      Iterate over alignment rows as SeqRecord objects.\n",
      "     |      \n",
      "     |      e.g.\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import IUPAC, Gapped\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
      "     |      >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
      "     |      >>> for record in align:\n",
      "     |      ...    print(record.id)\n",
      "     |      ...    print(record.seq)\n",
      "     |      ...\n",
      "     |      Alpha\n",
      "     |      ACTGCTAGCTAG\n",
      "     |      Beta\n",
      "     |      ACT-CTAGCTAG\n",
      "     |      Gamma\n",
      "     |      ACTGCTAGATAG\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |      Return the number of sequences in the alignment.\n",
      "     |      \n",
      "     |      Use len(alignment) to get the number of sequences (i.e. the number of\n",
      "     |      rows), and alignment.get_alignment_length() to get the length of the\n",
      "     |      longest sequence (i.e. the number of columns).\n",
      "     |      \n",
      "     |      This is easy to remember if you think of the alignment as being like a\n",
      "     |      list of SeqRecord objects.\n",
      "     |  \n",
      "     |  __repr__(self)\n",
      "     |      Return a representation of the object for debugging.\n",
      "     |      \n",
      "     |      The representation cannot be used with eval() to recreate the object,\n",
      "     |      which is usually possible with simple python ojects.  For example:\n",
      "     |      \n",
      "     |      <Bio.Align.MultipleSeqAlignment instance (2 records of length 14,\n",
      "     |      SingleLetterAlphabet()) at a3c184c>\n",
      "     |      \n",
      "     |      The hex string is the memory address of the object, see help(id).\n",
      "     |      This provides a simple way to visually distinguish alignments of\n",
      "     |      the same size.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Return a multi-line string summary of the alignment.\n",
      "     |      \n",
      "     |      This output is intended to be readable, but large alignments are\n",
      "     |      shown truncated.  A maximum of 20 rows (sequences) and 50 columns\n",
      "     |      are shown, with the record identifiers.  This should fit nicely on a\n",
      "     |      single screen. e.g.\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import IUPAC, Gapped\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
      "     |      >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
      "     |      >>> print(align)\n",
      "     |      Gapped(IUPACUnambiguousDNA(), '-') alignment with 3 rows and 12 columns\n",
      "     |      ACTGCTAGCTAG Alpha\n",
      "     |      ACT-CTAGCTAG Beta\n",
      "     |      ACTGCTAGATAG Gamma\n",
      "     |      \n",
      "     |      See also the alignment's format method.\n",
      "     |  \n",
      "     |  add_sequence(self, descriptor, sequence, start=None, end=None, weight=1.0)\n",
      "     |      Add a sequence to the alignment.\n",
      "     |      \n",
      "     |      This doesn't do any kind of alignment, it just adds in the sequence\n",
      "     |      object, which is assumed to be prealigned with the existing\n",
      "     |      sequences.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |          - descriptor - The descriptive id of the sequence being added.\n",
      "     |            This will be used as the resulting SeqRecord's\n",
      "     |            .id property (and, for historical compatibility,\n",
      "     |            also the .description property)\n",
      "     |          - sequence - A string with sequence info.\n",
      "     |          - start - You can explicitly set the start point of the sequence.\n",
      "     |            This is useful (at least) for BLAST alignments, which can\n",
      "     |            just be partial alignments of sequences.\n",
      "     |          - end - Specify the end of the sequence, which is important\n",
      "     |            for the same reason as the start.\n",
      "     |          - weight - The weight to place on the sequence in the alignment.\n",
      "     |            By default, all sequences have the same weight. (0.0 =>\n",
      "     |            no weight, 1.0 => highest weight)\n",
      "     |      \n",
      "     |      In general providing a SeqRecord and calling .append is preferred.\n",
      "     |  \n",
      "     |  append(self, record)\n",
      "     |      Add one more SeqRecord object to the alignment as a new row.\n",
      "     |      \n",
      "     |      This must have the same length as the original alignment (unless this is\n",
      "     |      the first record), and have an alphabet compatible with the alignment's\n",
      "     |      alphabet.\n",
      "     |      \n",
      "     |      >>> from Bio import AlignIO\n",
      "     |      >>> align = AlignIO.read(\"Clustalw/opuntia.aln\", \"clustal\")\n",
      "     |      >>> print(align)\n",
      "     |      SingleLetterAlphabet() alignment with 7 rows and 156 columns\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |      TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |      >>> len(align)\n",
      "     |      7\n",
      "     |      \n",
      "     |      We'll now construct a dummy record to append as an example:\n",
      "     |      \n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> dummy = SeqRecord(Seq(\"N\"*156), id=\"dummy\")\n",
      "     |      \n",
      "     |      Now append this to the alignment,\n",
      "     |      \n",
      "     |      >>> align.append(dummy)\n",
      "     |      >>> print(align)\n",
      "     |      SingleLetterAlphabet() alignment with 8 rows and 156 columns\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273285|gb|AF191659.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273284|gb|AF191658.1|AF191\n",
      "     |      TATACATTAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273287|gb|AF191661.1|AF191\n",
      "     |      TATACATAAAAGAAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273286|gb|AF191660.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273290|gb|AF191664.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273289|gb|AF191663.1|AF191\n",
      "     |      TATACATTAAAGGAGGGGGATGCGGATAAATGGAAAGGCGAAAG...AGA gi|6273291|gb|AF191665.1|AF191\n",
      "     |      NNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNNN...NNN dummy\n",
      "     |      >>> len(align)\n",
      "     |      8\n",
      "     |  \n",
      "     |  extend(self, records)\n",
      "     |      Add more SeqRecord objects to the alignment as rows.\n",
      "     |      \n",
      "     |      They must all have the same length as the original alignment, and have\n",
      "     |      alphabets compatible with the alignment's alphabet. For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import generic_dna\n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> a = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Alpha\")\n",
      "     |      >>> b = SeqRecord(Seq(\"AAA-CGT\", generic_dna), id=\"Beta\")\n",
      "     |      >>> c = SeqRecord(Seq(\"AAAAGGT\", generic_dna), id=\"Gamma\")\n",
      "     |      >>> d = SeqRecord(Seq(\"AAAACGT\", generic_dna), id=\"Delta\")\n",
      "     |      >>> e = SeqRecord(Seq(\"AAA-GGT\", generic_dna), id=\"Epsilon\")\n",
      "     |      \n",
      "     |      First we create a small alignment (three rows):\n",
      "     |      \n",
      "     |      >>> align = MultipleSeqAlignment([a, b, c])\n",
      "     |      >>> print(align)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      \n",
      "     |      Now we can extend this alignment with another two rows:\n",
      "     |      \n",
      "     |      >>> align.extend([d, e])\n",
      "     |      >>> print(align)\n",
      "     |      DNAAlphabet() alignment with 5 rows and 7 columns\n",
      "     |      AAAACGT Alpha\n",
      "     |      AAA-CGT Beta\n",
      "     |      AAAAGGT Gamma\n",
      "     |      AAAACGT Delta\n",
      "     |      AAA-GGT Epsilon\n",
      "     |      \n",
      "     |      Because the alignment object allows iteration over the rows as\n",
      "     |      SeqRecords, you can use the extend method with a second alignment\n",
      "     |      (provided its sequences have the same length as the original alignment).\n",
      "     |  \n",
      "     |  format(self, format)\n",
      "     |      Return the alignment as a string in the specified file format.\n",
      "     |      \n",
      "     |      The format should be a lower case string supported as an output\n",
      "     |      format by Bio.AlignIO (such as \"fasta\", \"clustal\", \"phylip\",\n",
      "     |      \"stockholm\", etc), which is used to turn the alignment into a\n",
      "     |      string.\n",
      "     |      \n",
      "     |      e.g.\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import IUPAC, Gapped\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
      "     |      >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
      "     |      >>> print(align.format(\"fasta\"))\n",
      "     |      >Alpha\n",
      "     |      ACTGCTAGCTAG\n",
      "     |      >Beta\n",
      "     |      ACT-CTAGCTAG\n",
      "     |      >Gamma\n",
      "     |      ACTGCTAGATAG\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> print(align.format(\"phylip\"))\n",
      "     |       3 12\n",
      "     |      Alpha      ACTGCTAGCT AG\n",
      "     |      Beta       ACT-CTAGCT AG\n",
      "     |      Gamma      ACTGCTAGAT AG\n",
      "     |      <BLANKLINE>\n",
      "     |      \n",
      "     |      For Python 2.6, 3.0 or later see also the built in format() function.\n",
      "     |  \n",
      "     |  get_alignment_length(self)\n",
      "     |      Return the maximum length of the alignment.\n",
      "     |      \n",
      "     |      All objects in the alignment should (hopefully) have the same\n",
      "     |      length. This function will go through and find this length\n",
      "     |      by finding the maximum length of sequences in the alignment.\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import IUPAC, Gapped\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> align = MultipleSeqAlignment([], Gapped(IUPAC.unambiguous_dna, \"-\"))\n",
      "     |      >>> align.add_sequence(\"Alpha\", \"ACTGCTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Beta\",  \"ACT-CTAGCTAG\")\n",
      "     |      >>> align.add_sequence(\"Gamma\", \"ACTGCTAGATAG\")\n",
      "     |      >>> align.get_alignment_length()\n",
      "     |      12\n",
      "     |      \n",
      "     |      If you want to know the number of sequences in the alignment,\n",
      "     |      use len(align) instead:\n",
      "     |      \n",
      "     |      >>> len(align)\n",
      "     |      3\n",
      "     |  \n",
      "     |  sort(self, key=None, reverse=False)\n",
      "     |      Sort the rows (SeqRecord objects) of the alignment in place.\n",
      "     |      \n",
      "     |      This sorts the rows alphabetically using the SeqRecord object id by\n",
      "     |      default. The sorting can be controlled by supplying a key function\n",
      "     |      which must map each SeqRecord to a sort value.\n",
      "     |      \n",
      "     |      This is useful if you want to add two alignments which use the same\n",
      "     |      record identifiers, but in a different order. For example,\n",
      "     |      \n",
      "     |      >>> from Bio.Alphabet import generic_dna\n",
      "     |      >>> from Bio.Seq import Seq\n",
      "     |      >>> from Bio.SeqRecord import SeqRecord\n",
      "     |      >>> from Bio.Align import MultipleSeqAlignment\n",
      "     |      >>> align1 = MultipleSeqAlignment([\n",
      "     |      ...              SeqRecord(Seq(\"ACGT\", generic_dna), id=\"Human\"),\n",
      "     |      ...              SeqRecord(Seq(\"ACGG\", generic_dna), id=\"Mouse\"),\n",
      "     |      ...              SeqRecord(Seq(\"ACGC\", generic_dna), id=\"Chicken\"),\n",
      "     |      ...          ])\n",
      "     |      >>> align2 = MultipleSeqAlignment([\n",
      "     |      ...              SeqRecord(Seq(\"CGGT\", generic_dna), id=\"Mouse\"),\n",
      "     |      ...              SeqRecord(Seq(\"CGTT\", generic_dna), id=\"Human\"),\n",
      "     |      ...              SeqRecord(Seq(\"CGCT\", generic_dna), id=\"Chicken\"),\n",
      "     |      ...          ])\n",
      "     |      \n",
      "     |      If you simple try and add these without sorting, you get this:\n",
      "     |      \n",
      "     |      >>> print(align1 + align2)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 8 columns\n",
      "     |      ACGTCGGT <unknown id>\n",
      "     |      ACGGCGTT <unknown id>\n",
      "     |      ACGCCGCT Chicken\n",
      "     |      \n",
      "     |      Consult the SeqRecord documentation which explains why you get a\n",
      "     |      default value when annotation like the identifier doesn't match up.\n",
      "     |      However, if we sort the alignments first, then add them we get the\n",
      "     |      desired result:\n",
      "     |      \n",
      "     |      >>> align1.sort()\n",
      "     |      >>> align2.sort()\n",
      "     |      >>> print(align1 + align2)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 8 columns\n",
      "     |      ACGCCGCT Chicken\n",
      "     |      ACGTCGTT Human\n",
      "     |      ACGGCGGT Mouse\n",
      "     |      \n",
      "     |      As an example using a different sort order, you could sort on the\n",
      "     |      GC content of each sequence.\n",
      "     |      \n",
      "     |      >>> from Bio.SeqUtils import GC\n",
      "     |      >>> print(align1)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 4 columns\n",
      "     |      ACGC Chicken\n",
      "     |      ACGT Human\n",
      "     |      ACGG Mouse\n",
      "     |      >>> align1.sort(key = lambda record: GC(record.seq))\n",
      "     |      >>> print(align1)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 4 columns\n",
      "     |      ACGT Human\n",
      "     |      ACGC Chicken\n",
      "     |      ACGG Mouse\n",
      "     |      \n",
      "     |      There is also a reverse argument, so if you wanted to sort by ID\n",
      "     |      but backwards:\n",
      "     |      \n",
      "     |      >>> align1.sort(reverse=True)\n",
      "     |      >>> print(align1)\n",
      "     |      DNAAlphabet() alignment with 3 rows and 4 columns\n",
      "     |      ACGG Mouse\n",
      "     |      ACGT Human\n",
      "     |      ACGC Chicken\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  column_annotations\n",
      "     |      Dictionary of per-letter-annotation for the sequence.\n",
      "    \n",
      "    class PairwiseAligner(_algorithms.PairwiseAligner)\n",
      "     |  Performs pairwise sequence alignment using dynamic programming.\n",
      "     |  \n",
      "     |  This provides functions to get global and local alignments between two\n",
      "     |  sequences.  A global alignment finds the best concordance between all\n",
      "     |  characters in two sequences.  A local alignment finds just the\n",
      "     |  subsequences that align the best.\n",
      "     |  \n",
      "     |  To perform a pairwise sequence alignment, first create a PairwiseAligner\n",
      "     |  object.  This object stores the match and mismatch scores, as well as the\n",
      "     |  gap scores.  Typically, match scores are positive, while mismatch scores\n",
      "     |  and gap scores are negative or zero.  By default, the match score is 1,\n",
      "     |  and the mismatch and gap scores are zero.  Based on the values of the gap\n",
      "     |  scores, a PairwiseAligner object automatically chooses the appropriate\n",
      "     |  alignment algorithm (the Needleman-Wunsch, Smith-Waterman, Gotoh, or\n",
      "     |  Waterman-Smith-Beyer global or local alignment algorithm).\n",
      "     |  \n",
      "     |  Calling the \"score\" method on the aligner with two sequences as arguments\n",
      "     |  will calculate the alignment score between the two sequences.\n",
      "     |  Calling the \"align\" method on the aligner with two sequences as arguments\n",
      "     |  will return a generator yielding the alignments between the two\n",
      "     |  sequences.\n",
      "     |  \n",
      "     |  Some examples:\n",
      "     |  \n",
      "     |  >>> from Bio import Align\n",
      "     |  >>> aligner = Align.PairwiseAligner()\n",
      "     |  >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
      "     |  >>> for alignment in sorted(alignments):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 3.0:\n",
      "     |  TACCG\n",
      "     |  -|-||\n",
      "     |  -A-CG\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 3.0:\n",
      "     |  TACCG\n",
      "     |  -||-|\n",
      "     |  -AC-G\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Specify the aligner mode as local to generate local alignments:\n",
      "     |  \n",
      "     |  >>> aligner.mode = 'local'\n",
      "     |  >>> alignments = aligner.align(\"TACCG\", \"ACG\")\n",
      "     |  >>> for alignment in sorted(alignments):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 3.0:\n",
      "     |  TACCG\n",
      "     |   |-||\n",
      "     |   A-CG\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 3.0:\n",
      "     |  TACCG\n",
      "     |   ||-|\n",
      "     |   AC-G\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Do a global alignment.  Identical characters are given 2 points,\n",
      "     |  1 point is deducted for each non-identical character.\n",
      "     |  \n",
      "     |  >>> aligner.mode = 'global'\n",
      "     |  >>> aligner.match_score = 2\n",
      "     |  >>> aligner.mismatch_score = -1\n",
      "     |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 6.0:\n",
      "     |  TACCG\n",
      "     |  -||-|\n",
      "     |  -AC-G\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 6.0:\n",
      "     |  TACCG\n",
      "     |  -|-||\n",
      "     |  -A-CG\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Same as above, except now 0.5 points are deducted when opening a\n",
      "     |  gap, and 0.1 points are deducted when extending it.\n",
      "     |  \n",
      "     |  >>> aligner.open_gap_score = -0.5\n",
      "     |  >>> aligner.extend_gap_score = -0.1\n",
      "     |  >>> aligner.target_end_gap_score = 0.0\n",
      "     |  >>> aligner.query_end_gap_score = 0.0\n",
      "     |  >>> for alignment in aligner.align(\"TACCG\", \"ACG\"):\n",
      "     |  ...     print(\"Score = %.1f:\" % alignment.score)\n",
      "     |  ...     print(alignment)\n",
      "     |  ...\n",
      "     |  Score = 5.5:\n",
      "     |  TACCG\n",
      "     |  -|-||\n",
      "     |  -A-CG\n",
      "     |  <BLANKLINE>\n",
      "     |  Score = 5.5:\n",
      "     |  TACCG\n",
      "     |  -||-|\n",
      "     |  -AC-G\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  The alignment function can also use known matrices already included in\n",
      "     |  Biopython:\n",
      "     |  \n",
      "     |  >>> from Bio.Align import substitution_matrices\n",
      "     |  >>> aligner = Align.PairwiseAligner()\n",
      "     |  >>> aligner.substitution_matrix = substitution_matrices.load(\"BLOSUM62\")\n",
      "     |  >>> alignments = aligner.align(\"KEVLA\", \"EVL\")\n",
      "     |  >>> alignments = list(alignments)\n",
      "     |  >>> print(\"Number of alignments: %d\" % len(alignments))\n",
      "     |  Number of alignments: 1\n",
      "     |  >>> alignment = alignments[0]\n",
      "     |  >>> print(\"Score = %.1f\" % alignment.score)\n",
      "     |  Score = 13.0\n",
      "     |  >>> print(alignment)\n",
      "     |  KEVLA\n",
      "     |  -|||-\n",
      "     |  -EVL-\n",
      "     |  <BLANKLINE>\n",
      "     |  \n",
      "     |  Method resolution order:\n",
      "     |      PairwiseAligner\n",
      "     |      _algorithms.PairwiseAligner\n",
      "     |      builtins.object\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __setattr__(self, key, value)\n",
      "     |      Implement setattr(self, name, value).\n",
      "     |  \n",
      "     |  align(self, seqA, seqB)\n",
      "     |      Return the alignments of two sequences using PairwiseAligner.\n",
      "     |  \n",
      "     |  score(self, seqA, seqB)\n",
      "     |      Return the alignments score of two sequences using PairwiseAligner.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Methods inherited from _algorithms.PairwiseAligner:\n",
      "     |  \n",
      "     |  __init__(self, /, *args, **kwargs)\n",
      "     |      Initialize self.  See help(type(self)) for accurate signature.\n",
      "     |  \n",
      "     |  __repr__(self, /)\n",
      "     |      Return repr(self).\n",
      "     |  \n",
      "     |  __str__(self, /)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Static methods inherited from _algorithms.PairwiseAligner:\n",
      "     |  \n",
      "     |  __new__(*args, **kwargs) from builtins.type\n",
      "     |      Create and return a new object.  See help(type) for accurate signature.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors inherited from _algorithms.PairwiseAligner:\n",
      "     |  \n",
      "     |  algorithm\n",
      "     |      alignment algorithm\n",
      "     |  \n",
      "     |  alphabet\n",
      "     |      alphabet\n",
      "     |  \n",
      "     |  end_extend_gap_score\n",
      "     |      end extend gap score\n",
      "     |  \n",
      "     |  end_gap_score\n",
      "     |      end gap score\n",
      "     |  \n",
      "     |  end_open_gap_score\n",
      "     |      end open gap score\n",
      "     |  \n",
      "     |  epsilon\n",
      "     |      roundoff epsilon\n",
      "     |  \n",
      "     |  extend_gap_score\n",
      "     |      extend gap score\n",
      "     |  \n",
      "     |  gap_score\n",
      "     |      gap score\n",
      "     |  \n",
      "     |  internal_extend_gap_score\n",
      "     |      internal extend gap score\n",
      "     |  \n",
      "     |  internal_gap_score\n",
      "     |      internal gap score\n",
      "     |  \n",
      "     |  internal_open_gap_score\n",
      "     |      internal open gap score\n",
      "     |  \n",
      "     |  left_extend_gap_score\n",
      "     |      left extend gap score\n",
      "     |  \n",
      "     |  left_gap_score\n",
      "     |      left gap score\n",
      "     |  \n",
      "     |  left_open_gap_score\n",
      "     |      left open gap score\n",
      "     |  \n",
      "     |  match\n",
      "     |      match score\n",
      "     |  \n",
      "     |  match_score\n",
      "     |      match score\n",
      "     |  \n",
      "     |  mismatch\n",
      "     |      mismatch score\n",
      "     |  \n",
      "     |  mismatch_score\n",
      "     |      mismatch score\n",
      "     |  \n",
      "     |  mode\n",
      "     |      alignment mode ('global' or 'local')\n",
      "     |  \n",
      "     |  open_gap_score\n",
      "     |      internal and end open gap score\n",
      "     |  \n",
      "     |  query_end_extend_gap_score\n",
      "     |      query end extend score\n",
      "     |  \n",
      "     |  query_end_gap_score\n",
      "     |      query end score\n",
      "     |  \n",
      "     |  query_end_open_gap_score\n",
      "     |      query end open score\n",
      "     |  \n",
      "     |  query_extend_gap_score\n",
      "     |      query gap extend score\n",
      "     |  \n",
      "     |  query_gap_score\n",
      "     |      query gap score\n",
      "     |  \n",
      "     |  query_internal_extend_gap_score\n",
      "     |      query internal extend gap score\n",
      "     |  \n",
      "     |  query_internal_gap_score\n",
      "     |      query internal gap score\n",
      "     |  \n",
      "     |  query_internal_open_gap_score\n",
      "     |      query internal open gap score\n",
      "     |  \n",
      "     |  query_left_extend_gap_score\n",
      "     |      query left extend score\n",
      "     |  \n",
      "     |  query_left_gap_score\n",
      "     |      query left score\n",
      "     |  \n",
      "     |  query_left_open_gap_score\n",
      "     |      query left open score\n",
      "     |  \n",
      "     |  query_open_gap_score\n",
      "     |      query gap open score\n",
      "     |  \n",
      "     |  query_right_extend_gap_score\n",
      "     |      query right extend score\n",
      "     |  \n",
      "     |  query_right_gap_score\n",
      "     |      query right score\n",
      "     |  \n",
      "     |  query_right_open_gap_score\n",
      "     |      query right open score\n",
      "     |  \n",
      "     |  right_extend_gap_score\n",
      "     |      right extend gap score\n",
      "     |  \n",
      "     |  right_gap_score\n",
      "     |      right gap score\n",
      "     |  \n",
      "     |  right_open_gap_score\n",
      "     |      right open gap score\n",
      "     |  \n",
      "     |  substitution_matrix\n",
      "     |      substitution_matrix\n",
      "     |  \n",
      "     |  target_end_extend_gap_score\n",
      "     |      target end extend gap score\n",
      "     |  \n",
      "     |  target_end_gap_score\n",
      "     |      target end gap score\n",
      "     |  \n",
      "     |  target_end_open_gap_score\n",
      "     |      target end open gap score\n",
      "     |  \n",
      "     |  target_extend_gap_score\n",
      "     |      target extend gap score\n",
      "     |  \n",
      "     |  target_gap_score\n",
      "     |      target gap score\n",
      "     |  \n",
      "     |  target_internal_extend_gap_score\n",
      "     |      target internal extend gap score\n",
      "     |  \n",
      "     |  target_internal_gap_score\n",
      "     |      target internal gap score\n",
      "     |  \n",
      "     |  target_internal_open_gap_score\n",
      "     |      target internal open gap score\n",
      "     |  \n",
      "     |  target_left_extend_gap_score\n",
      "     |      target left extend score\n",
      "     |  \n",
      "     |  target_left_gap_score\n",
      "     |      target left score\n",
      "     |  \n",
      "     |  target_left_open_gap_score\n",
      "     |      target left open score\n",
      "     |  \n",
      "     |  target_open_gap_score\n",
      "     |      target open gap score\n",
      "     |  \n",
      "     |  target_right_extend_gap_score\n",
      "     |      target right extend score\n",
      "     |  \n",
      "     |  target_right_gap_score\n",
      "     |      target right score\n",
      "     |  \n",
      "     |  target_right_open_gap_score\n",
      "     |      target right open score\n",
      "    \n",
      "    class PairwiseAlignment(builtins.object)\n",
      "     |  PairwiseAlignment(target, query, path, score)\n",
      "     |  \n",
      "     |  Represents a pairwise sequence alignment.\n",
      "     |  \n",
      "     |  Internally, the pairwise alignment is stored as the path through\n",
      "     |  the traceback matrix, i.e. a tuple of pairs of indices corresponding\n",
      "     |  to the vertices of the path in the traceback matrix.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __cmp__(self, other)\n",
      "     |      # For Python2 only\n",
      "     |  \n",
      "     |  __eq__(self, other)\n",
      "     |      Return self==value.\n",
      "     |  \n",
      "     |  __format__(self, format_spec)\n",
      "     |      Default object formatter.\n",
      "     |  \n",
      "     |  __ge__(self, other)\n",
      "     |      Return self>=value.\n",
      "     |  \n",
      "     |  __gt__(self, other)\n",
      "     |      Return self>value.\n",
      "     |  \n",
      "     |  __init__(self, target, query, path, score)\n",
      "     |      Initialize a new PairwiseAlignment object.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - target  - The first sequence, as a plain string, without gaps.\n",
      "     |       - query   - The second sequence, as a plain string, without gaps.\n",
      "     |       - path    - The path through the traceback matrix, defining an\n",
      "     |                   alignment.\n",
      "     |       - score   - The alignment score.\n",
      "     |      \n",
      "     |      You would normally obtain a PairwiseAlignment object by iterating\n",
      "     |      over a PairwiseAlignments object.\n",
      "     |  \n",
      "     |  __le__(self, other)\n",
      "     |      Return self<=value.\n",
      "     |  \n",
      "     |  __lt__(self, other)\n",
      "     |      Return self<value.\n",
      "     |  \n",
      "     |  __ne__(self, other)\n",
      "     |      Return self!=value.\n",
      "     |  \n",
      "     |  __str__(self)\n",
      "     |      Return str(self).\n",
      "     |  \n",
      "     |  format(self)\n",
      "     |      Create a human-readable representation of the alignment.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "     |  \n",
      "     |  aligned\n",
      "     |      Return the indices of subsequences aligned to each other.\n",
      "     |      \n",
      "     |      This property returns the start and end indices of subsequences\n",
      "     |      in the target and query sequence that were aligned to each other.\n",
      "     |      If the alignment between target (t) and query (q) consists of N\n",
      "     |      chunks, you get two tuples of length N:\n",
      "     |      \n",
      "     |          (((t_start1, t_end1), (t_start2, t_end2), ..., (t_startN, t_endN)),\n",
      "     |           ((q_start1, q_end1), (q_start2, q_end2), ..., (q_startN, q_endN)))\n",
      "     |      \n",
      "     |      For example,\n",
      "     |      \n",
      "     |      >>> from Bio import Align\n",
      "     |      >>> aligner = Align.PairwiseAligner()\n",
      "     |      >>> alignments = aligner.align(\"GAACT\", \"GAT\")\n",
      "     |      >>> alignment = alignments[0]\n",
      "     |      >>> print(alignment)\n",
      "     |      GAACT\n",
      "     |      ||--|\n",
      "     |      GA--T\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.aligned\n",
      "     |      (((0, 2), (4, 5)), ((0, 2), (2, 3)))\n",
      "     |      >>> alignment = alignments[1]\n",
      "     |      >>> print(alignment)\n",
      "     |      GAACT\n",
      "     |      |-|-|\n",
      "     |      G-A-T\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignment.aligned\n",
      "     |      (((0, 1), (2, 3), (4, 5)), ((0, 1), (1, 2), (2, 3)))\n",
      "     |      \n",
      "     |      Note that different alignments may have the same subsequences\n",
      "     |      aligned to each other. In particular, this may occur if alignments\n",
      "     |      differ from each other in terms of their gap placement only:\n",
      "     |      \n",
      "     |      >>> aligner.mismatch_score = -10\n",
      "     |      >>> alignments = aligner.align(\"AAACAAA\", \"AAAGAAA\")\n",
      "     |      >>> len(alignments)\n",
      "     |      2\n",
      "     |      >>> print(alignments[0])\n",
      "     |      AAAC-AAA\n",
      "     |      |||--|||\n",
      "     |      AAA-GAAA\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignments[0].aligned\n",
      "     |      (((0, 3), (4, 7)), ((0, 3), (4, 7)))\n",
      "     |      >>> print(alignments[1])\n",
      "     |      AAA-CAAA\n",
      "     |      |||--|||\n",
      "     |      AAAG-AAA\n",
      "     |      <BLANKLINE>\n",
      "     |      >>> alignments[1].aligned\n",
      "     |      (((0, 3), (4, 7)), ((0, 3), (4, 7)))\n",
      "     |      \n",
      "     |      The property can be used to identify alignments that are identical\n",
      "     |      to each other in terms of their aligned sequences.\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data and other attributes defined here:\n",
      "     |  \n",
      "     |  __hash__ = None\n",
      "    \n",
      "    class PairwiseAlignments(builtins.object)\n",
      "     |  PairwiseAlignments(seqA, seqB, score, paths)\n",
      "     |  \n",
      "     |  Implements an iterator over pairwise alignments returned by the aligner.\n",
      "     |  \n",
      "     |  This class also supports indexing, which is fast for increasing indices,\n",
      "     |  but may be slow for random access of a large number of alignments.\n",
      "     |  \n",
      "     |  Note that pairwise aligners can return an astronomical number of alignments,\n",
      "     |  even for relatively short sequences, if they align poorly to each other. We\n",
      "     |  therefore recommend to first check the number of alignments, accessible as\n",
      "     |  len(alignments), which can be calculated quickly even if the number of\n",
      "     |  alignments is very large.\n",
      "     |  \n",
      "     |  Methods defined here:\n",
      "     |  \n",
      "     |  __getitem__(self, index)\n",
      "     |  \n",
      "     |  __init__(self, seqA, seqB, score, paths)\n",
      "     |      Initialize a new PairwiseAlignments object.\n",
      "     |      \n",
      "     |      Arguments:\n",
      "     |       - seqA  - The first sequence, as a plain string, without gaps.\n",
      "     |       - seqB  - The second sequence, as a plain string, without gaps.\n",
      "     |       - score - The alignment score.\n",
      "     |       - paths - An iterator over the paths in the traceback matrix;\n",
      "     |                 each path defines one alignment.\n",
      "     |      \n",
      "     |      You would normally obtain an PairwiseAlignments object by calling\n",
      "     |      aligner.align(seqA, seqB), where aligner is a PairwiseAligner object.\n",
      "     |  \n",
      "     |  __iter__(self)\n",
      "     |  \n",
      "     |  __len__(self)\n",
      "     |  \n",
      "     |  __next__(self)\n",
      "     |  \n",
      "     |  ----------------------------------------------------------------------\n",
      "     |  Data descriptors defined here:\n",
      "     |  \n",
      "     |  __dict__\n",
      "     |      dictionary for instance variables (if defined)\n",
      "     |  \n",
      "     |  __weakref__\n",
      "     |      list of weak references to the object (if defined)\n",
      "\n",
      "DATA\n",
      "    print_function = _Feature((2, 6, 0, 'alpha', 2), (3, 0, 0, 'alpha', 0)...\n",
      "\n",
      "FILE\n",
      "    d:\\anaconda\\lib\\site-packages\\bio\\align\\__init__.py\n",
      "\n",
      "\n"
     ]
    }
   ],
   "source": [
    "help(Align)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "A-CACACTA\n",
      "| ||||| |\n",
      "AGCACAC-A\n",
      "  Score=7\n",
      "\n"
     ]
    }
   ],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  },
  "latex_envs": {
   "LaTeX_envs_menu_present": true,
   "autoclose": false,
   "autocomplete": true,
   "bibliofile": "biblio.bib",
   "cite_by": "apalike",
   "current_citInitial": 1,
   "eqLabelWithNumbers": true,
   "eqNumInitial": 1,
   "hotkeys": {
    "equation": "Ctrl-E",
    "itemize": "Ctrl-I"
   },
   "labels_anchors": false,
   "latex_user_defs": false,
   "report_style_numbering": false,
   "user_envs_cfg": false
  },
  "toc": {
   "base_numbering": 1,
   "nav_menu": {},
   "number_sections": true,
   "sideBar": true,
   "skip_h1_title": false,
   "title_cell": "Table of Contents",
   "title_sidebar": "Contents",
   "toc_cell": false,
   "toc_position": {},
   "toc_section_display": true,
   "toc_window_display": false
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
